examples/mgpcg_advanced.py

```diff
@@ -197,12 +237,48 @@ def run(self):
             # self.beta = new_rTr / old_rTr
             self.reduce(self.z[0], self.r[0])
             new_zTr = self.sum[None]
-            self.beta[None] = new_zTr / old_zTr
+            self.beta[None] = new_zTr / (old_zTr + eps)
 
             # self.p = self.z + self.beta self.p
             self.update_p()
             old_zTr = new_zTr
 
+            yield rTr
```

Comment:
Usually people just iterate until the residual is below a threshold. No need for a generator here.